### PR TITLE
Add some Utilities for Pipeline Surgery

### PR DESCRIPTION
**Pull request by tgroh**
Be sure to do all of the following to help us incorporate your contribution
quickly and easily:

 - [x] Make sure the PR title is formatted like:
   `[BEAM-<Jira issue #>] Description of pull request`
 - [x] Make sure tests pass via `mvn clean verify`. (Even better, enable
       Travis-CI on your fork and ensure the whole test matrix passes).
 - [x] Replace `<Jira issue #>` in the title with the actual Jira issue
       number, if there is one.
 - [x] If this contribution is large, please file an Apache
       [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.txt).

---
Add some additional matchers and Override Factories to runners-core-construction

These are going to be used within the `DataflowRunner`

### CODE REVIEW

**Comment by tgroh** [comment]
R: @kennknowles

File: runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/PTransformMatchers.java [hunk 1]
```diff
@@ -158,4 +160,39 @@ public boolean matches(AppliedPTransform<?, ?, ?> application) {
...
+   * A {@link PTransformMatcher} which matches a {@link ParDo.Bound} or {@link ParDo.BoundMulti}
+   * where the {@link DoFn} is of the provided type.
+   */
+  public static PTransformMatcher parDoWithFnType(final Class<? extends DoFn> fnType) {
+    return new PTransformMatcher() {
+      @Override
+      public boolean matches(AppliedPTransform<?, ?, ?> application) {
+        DoFn<?, ?> fn;
+        if (application.getTransform() instanceof ParDo.Bound) {
+          fn = ((ParDo.Bound) application.getTransform()).getFn();
```
Comment (kennknowles) [inline runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/PTransformMatchers.java]: This code basically does this, but I just want to make sure that your intent matches how it should work: In a world where this matcher is running over a reconstructed graph, the matcher just fails to match anything that isn't deserialized into known Java-SDK-specific classes, and also fails to match if the DoFn is junk from another language.

File: runners/core-construction-java/src/test/java/org/apache/beam/runners/core/construction/PTransformMatchersTest.java [hunk 2]
```diff
@@ -270,4 +276,107 @@ public void parDoMultiWithTimers() {
...
+      public void process(ProcessContext ctxt) {
+      }
+    };
+    AppliedPTransform<?, ?, ?> parDoSingle = getAppliedTransform(ParDo.of(fn));
+    AppliedPTransform<?, ?, ?> parDoMulti =
+        getAppliedTransform(
+            ParDo.of(fn).withOutputTags(new TupleTag<Object>(), TupleTagList.empty()));
+
+    PTransformMatcher matcher = PTransformMatchers.parDoWithFnType(doFnWithState.getClass());
+    assertThat(matcher.matches(parDoSingle), is(false));
```
Comment (kennknowles) [inline runners/core-construction-java/src/test/java/org/apache/beam/runners/core/construction/PTransformMatchersTest.java]: Eventually we might need a way for a `ParDo` to exist that does not have a valid Java DoFn _or_ have a `PTransform` with a URN of `urn:beam:transform:pardo` that is not an instance of `ParDo.Bound`, etc. Just a heads up. We aren't there yet, and the problem will be its own forcing function when we get there.

File: runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/PTransformMatchers.java (hunk 1 above)
Comment (tgroh) [inline runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/PTransformMatchers.java]: That's accurate. Theoretically for well-known DoFns with a useful URN a similar matcher could be constructed

File: runners/core-construction-java/src/test/java/org/apache/beam/runners/core/construction/PTransformMatchersTest.java (hunk 2 above)
Comment (tgroh) [inline runners/core-construction-java/src/test/java/org/apache/beam/runners/core/construction/PTransformMatchersTest.java]: Yeah, that's on the critical path for supporting URNs in general I believe.

**Pull request closed without merging**